phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,456 @@
+        }
+    }
+
+    private static void checkAndWaitForMajorCompaction(Admin admin, Map.Entry<String, ArrayList<String>> value) {
+        value.getValue().parallelStream().forEach(table -> {
+            try {
+                CompactionState cs = admin.getCompactionState(TableName.valueOf(table));
+                while (cs.equals(CompactionState.MAJOR)) {
+                    LOGGER.fine("Major compaction state is " + cs.name() +" for "+table+" sleeping for 5 secs before next check");
+                    if (!dryRun) {
+                        Thread.sleep(5000);
+                        cs = admin.getCompactionState(TableName.valueOf(table));
+                    }
+                }
+                if (!dryRun) {
+                    LOGGER.fine("Major compaction state is " + cs.name() + " for " + table);
+                }
+            } catch (InterruptedException | IOException e) {
+                LOGGER.severe("Something went wrong while checking and waiting for Major Compaction for table " + table+e);
+            }
+        });
+    }
+
+    private static void resetMajorCompaction(Admin admin, Map.Entry<String, ArrayList<String>> value) {
+        value.getValue().parallelStream().forEach(table -> {
+            try {
+                tableMC.put(table, admin.getDescriptor(TableName.valueOf(table)).isCompactionEnabled());
+                if (!dryRun) {
+                    TableDescriptorBuilder tdb = TableDescriptorBuilder.newBuilder(admin.getDescriptor(TableName.valueOf(table)));
+                    tdb.setCompactionEnabled(false);
+                    admin.modifyTable(tdb.build());
+                }
+                LOGGER.info("Major compaction disabled for "+table);
+
+            } catch (IOException e) {
+                LOGGER.severe("Something went wrong while setting Major Compaction to false for "+table+" "+e);
+            }
+        });
+    }
+
+    private static void setMajorCompaction(Admin admin, Map.Entry<String, ArrayList<String>> value) {
+        value.getValue().parallelStream().forEach(table -> {
+            try {
+                if (!dryRun) {
+                    TableDescriptorBuilder tdb = TableDescriptorBuilder.newBuilder(admin.getDescriptor(TableName.valueOf(table)));
+                    tdb.setCompactionEnabled(tableMC.get(table));
+                    admin.modifyTable(tdb.build());
+                }
+                LOGGER.info("Major compaction set to original value {"+tableMC.get(table)+"} for "+table);
+            } catch (IOException e) {
+                LOGGER.severe("Something went wrong while setting Major Compaction to old value "+table+" "+e);
+            }
+        });
+    }
+
+    private static void rebuildIndexes(Map.Entry<String, ArrayList<String>> value, Configuration conf) {
+        String schema = value.getKey().split("\\.")[0];
+        String table = value.getKey().split("\\.")[1];
+
+        value.getValue().stream().forEach(index -> {
```

Comment:
@swaroopak - unfortunately, we don't allow JDK 8 and up features like stream() in Phoenix 4.x, because HBase 1.x doesn't allow them. Since this PR is bound for master, which is 5.x, it's technically OK here, but you'd have to rewrite this for the backport, which is probably more trouble than we want to maintain.

Reply:
Okay. I ll discontinue using lambda function.